From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yuntian Zhang <yt@radxa.com>
Date: Mon, 27 Jun 2022 14:24:38 +0800
Subject: [PATCH] mmc: meson_gx_mmc: Add MMC_MESON_GX_HS52_MODE config

This option can be turned off to disable MMC High Speed 52MHz on the
Meson controller. This helps eMMC stability on some faulty boards.

Signed-off-by: Yuntian Zhang <yt@radxa.com>
---
 drivers/mmc/Kconfig        | 9 +++++++++
 drivers/mmc/meson_gx_mmc.c | 4 ++++
 2 files changed, 13 insertions(+)

diff --git a/drivers/mmc/Kconfig b/drivers/mmc/Kconfig
index f04cc44e19..30f276749e 100644
--- a/drivers/mmc/Kconfig
+++ b/drivers/mmc/Kconfig
@@ -286,6 +286,15 @@ config MMC_MESON_GX
 	help
 	 Support for EMMC host controller on Meson GX ARM SoCs platform (S905)
 
+config MMC_MESON_GX_HS52_MODE
+	bool "Enable MMC High Speed 52MHz mode for Meson GX EMMC controller"
+	depends on MMC_MESON_GX
+	default y
+	help
+	 This selects HS52 mode for Meson controller. Some boards have stability
+	 issue when eMMC is running at this speed, and turning HS52 off allows
+	 eMMC to function properly.
+
 config MMC_MXC
 	bool "Freescale i.MX21/27/31 or MPC512x Multimedia Card support"
 	help
diff --git a/drivers/mmc/meson_gx_mmc.c b/drivers/mmc/meson_gx_mmc.c
index fcf4f03d1e..3385acfe64 100644
--- a/drivers/mmc/meson_gx_mmc.c
+++ b/drivers/mmc/meson_gx_mmc.c
@@ -277,7 +277,11 @@ static int meson_mmc_probe(struct udevice *dev)
 	cfg->voltages = MMC_VDD_33_34 | MMC_VDD_32_33 |
 			MMC_VDD_31_32 | MMC_VDD_165_195;
 	cfg->host_caps = MMC_MODE_8BIT | MMC_MODE_4BIT |
+#ifdef CONFIG_MMC_MESON_GX_HS52_MODE
 			MMC_MODE_HS_52MHz | MMC_MODE_HS;
+#else
+			MMC_MODE_HS;
+#endif
 	cfg->f_min = DIV_ROUND_UP(SD_EMMC_CLKSRC_24M, CLK_MAX_DIV);
 	cfg->f_max = 100000000; /* 100 MHz */
 	cfg->b_max = 511; /* max 512 - 1 blocks */
-- 
2.36.1

